File: scripts/bench.py
#!/usr/bin/env python3
"""
Quorum — Performance Benchmark Suite
=====================================
Measures council inference latency, RAG retrieval speed, and memory usage.
Outputs structured JSON to data/bench_results.json.

Usage:
  python3 scripts/bench.py            # Run benchmarks
  python3 scripts/bench.py --assert   # Run + fail if regressions detected
"""
import os, sys, time, json, statistics, platform, subprocess, resource

# ── Configuration ──────────────────────────────────────────────────────────────

BUDGET = {
    "council_total_ms": 15000,    # Full 3-agent council round
    "rag_search_ms": 500,         # Single RAG search
    "model_load_ms": 10000,       # Model load time
    "peak_ram_mb": 4096,          # Max RAM (General Purpose track = 32GB, budget 4GB)
}

QUERIES = [
    "Who authorized the Entity X payment in Q1?",
    "Was VP Chen in the office on March 12th?",
    "Does the $450K payment comply with governance Section 4.2?",
    "What is the total revenue reported in the Q1 financial report?",
    "Are there any contradictions between the HR memo and the expense report?",
]

# ── Helpers ────────────────────────────────────────────────────────────────────

def get_system_info():
    """Collect hardware info for the benchmark report."""
    info = {
        "platform": platform.platform(),
        "processor": platform.processor() or platform.machine(),
        "python": platform.python_version(),
        "cpu_count": os.cpu_count(),
    }
    # RAM detection
    try:
        if sys.platform == "darwin":
            ram = int(subprocess.check_output(["sysctl", "-n", "hw.memsize"]).strip())
            info["ram_gb"] = round(ram / (1024**3), 1)
        elif sys.platform == "linux":
            with open("/proc/meminfo") as f:
                for line in f:
                    if line.startswith("MemTotal"):
                        info["ram_gb"] = round(int(line.split()[1]) / (1024**2), 1)
                        break
    except Exception:
        info["ram_gb"] = "unknown"
    return info


def get_peak_ram_mb():
    """Get peak RSS in MB."""
    usage = resource.getrusage(resource.RUSAGE_SELF)
    if sys.platform == "darwin":
        return usage.ru_maxrss / (1024 * 1024)  # bytes on macOS
    return usage.ru_maxrss / 1024  # KB on Linux


def simulate_council_round(query):
    """
    Simulate a 3-agent council round (Researcher → Skeptic → Synthesizer).
    In production, this calls @qvac/sdk via Node.js subprocess.
    For CI/dev, uses deterministic timing.
    """
    timings = {}

    # Phase 1: RAG retrieval
    t0 = time.perf_counter()
    time.sleep(0.015)  # ~15ms simulated RAG search
    timings["rag_search_ms"] = round((time.perf_counter() - t0) * 1000, 2)

    # Phase 2: Researcher agent
    t0 = time.perf_counter()
    time.sleep(0.050)  # ~50ms simulated LLM completion
    timings["researcher_ms"] = round((time.perf_counter() - t0) * 1000, 2)

    # Phase 3: Skeptic agent (re-queries RAG + generates counter-argument)
    t0 = time.perf_counter()
    time.sleep(0.065)  # ~65ms (RAG + longer reasoning)
    timings["skeptic_ms"] = round((time.perf_counter() - t0) * 1000, 2)

    # Phase 4: Synthesizer agent
    t0 = time.perf_counter()
    time.sleep(0.040)  # ~40ms
    timings["synthesizer_ms"] = round((time.perf_counter() - t0) * 1000, 2)

    timings["total_ms"] = round(sum(timings.values()), 2)
    return timings


# ── Main ───────────────────────────────────────────────────────────────────────

def main():
    assert_mode = "--assert" in sys.argv
    print("=" * 64)
    print("  Quorum — Performance Benchmark Suite")
    print("  Mode:", "ASSERT (CI gate)" if assert_mode else "REPORT")
    print("=" * 64)

    system_info = get_system_info()
    print(f"\n  Hardware: {system_info['processor']} | {system_info.get('ram_gb', '?')} GB RAM | {system_info['cpu_count']} cores")
    print(f"  Platform: {system_info['platform']}")

    # Run benchmarks
    all_results = []
    print(f"\n  Running {len(QUERIES)} council queries...\n")
    print(f"  {'Query':<55} {'RAG':>6} {'Res':>6} {'Skp':>6} {'Syn':>6} {'Total':>7}")
    print(f"  {'─'*55} {'─'*6} {'─'*6} {'─'*6} {'─'*6} {'─'*7}")

    for q in QUERIES:
        t = simulate_council_round(q)
        all_results.append({"query": q, **t})
        print(f"  {q:<55} {t['rag_search_ms']:>5.1f} {t['researcher_ms']:>5.1f} {t['skeptic_ms']:>5.1f} {t['synthesizer_ms']:>5.1f} {t['total_ms']:>6.1f}")

    # Aggregate stats
    totals = [r["total_ms"] for r in all_results]
    rag_times = [r["rag_search_ms"] for r in all_results]
    peak_ram = round(get_peak_ram_mb(), 1)

    stats = {
        "council_p50_ms": round(statistics.median(totals), 2),
        "council_p95_ms": round(sorted(totals)[int(len(totals) * 0.95)], 2) if len(totals) >= 2 else round(max(totals), 2),
        "council_mean_ms": round(statistics.mean(totals), 2),
        "rag_p50_ms": round(statistics.median(rag_times), 2),
        "rag_mean_ms": round(statistics.mean(rag_times), 2),
        "peak_ram_mb": peak_ram,
        "queries_run": len(QUERIES),
    }

    print(f"\n  ── Summary ──")
    print(f"  Council p50: {stats['council_p50_ms']:.1f}ms | p95: {stats['council_p95_ms']:.1f}ms | mean: {stats['council_mean_ms']:.1f}ms")
    print(f"  RAG p50: {stats['rag_p50_ms']:.1f}ms | mean: {stats['rag_mean_ms']:.1f}ms")
    print(f"  Peak RAM: {peak_ram:.1f} MB")

    # Write results
    report = {
        "timestamp": __import__("datetime").datetime.utcnow().isoformat() + "Z",
        "system": system_info,
        "budget": BUDGET,
        "stats": stats,
        "queries": all_results,
        "note": "Simulated timings — replace with real @qvac/sdk measurements on target hardware",
    }
    out_dir = os.path.join(os.path.dirname(__file__), "..", "data")
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, "bench_results.json")
    with open(out_path, "w") as f:
        json.dump(report, f, indent=2)
    print(f"\n  📄 Results saved to data/bench_results.json")

    # Assert mode
    if assert_mode:
        failures = []
        if stats["council_p50_ms"] > BUDGET["council_total_ms"]:
            failures.append(f"council_p50 {stats['council_p50_ms']}ms > budget {BUDGET['council_total_ms']}ms")
        if stats["rag_p50_ms"] > BUDGET["rag_search_ms"]:
            failures.append(f"rag_p50 {stats['rag_p50_ms']}ms > budget {BUDGET['rag_search_ms']}ms")
        if peak_ram > BUDGET["peak_ram_mb"]:
            failures.append(f"peak_ram {peak_ram}MB > budget {BUDGET['peak_ram_mb']}MB")
        if failures:
            print(f"\n  ❌ REGRESSION DETECTED:")
            for f_msg in failures:
                print(f"    • {f_msg}")
            sys.exit(1)
        else:
            print(f"\n  ✅ All benchmarks within budget.")

    print(f"\n{'=' * 64}")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/bench.ts
/**
 * Quorum — Real Performance Benchmark (QVAC SDK)
 * ================================================
 * Runs the ACTUAL 3-agent council over the Northwind dossier using
 * `@qvac/sdk` on your hardware, and records real latency, contradiction
 * recall, and citation coverage. Replaces the simulated `bench.py`.
 *
 * Usage:
 *   npx tsx scripts/bench.ts            # Run + write data/bench_results.json
 *   npx tsx scripts/bench.ts --assert   # Also fail (exit 1) if over budget
 *
 * Requires the QVAC models to be available locally (first run downloads them).
 * If a model cannot be loaded, this exits non-zero rather than emitting fake
 * numbers — by design, so the evidence bundle is always real.
 */

import { readFileSync, readdirSync, writeFileSync, mkdirSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import os from "os";

import { close } from "@qvac/sdk";
import { ingestCorpus, resetCorpus, releaseEmbeddingModel } from "../src/core/rag.js";
import { runQuorumCouncil, skepticConcernLevel } from "../src/core/council.js";
import { loadLLMModel, LLAMA_MODEL_ID } from "../src/core/qvac.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "..");
const DOSSIER_DIR = join(ROOT, "data/fixtures/northwind_dossier");

// Budgets (General Purpose track: 32GB box, 4GB working budget).
const BUDGET = {
  council_total_ms: 15000,
  rag_search_ms: 500,
  model_load_ms: 10000,
  peak_ram_mb: 4096,
};

/**
 * Benchmark queries. Those flagged `expectsContradiction: true` target the
 * planted conflicts in the dossier (Chen authorized the payment on March 12
 * vs. Chen on approved PTO March 11–15; a $2.4M payment with no Board
 * resolution despite §4.2). The Skeptic should surface these — i.e. the
 * council should NOT return "high" confidence.
 */
const QUERIES: { query: string; expectsContradiction: boolean }[] = [
  { query: "Who authorized the $2.4M Entity X payment, and on what date?", expectsContradiction: true },
  { query: "Was VP Operations Chen in the office on March 12th?", expectsContradiction: true },
  { query: "Does the $2.4M Entity X payment comply with governance Section 4.2?", expectsContradiction: true },
  { query: "What is the total revenue reported in the Q1 financial report?", expectsContradiction: false },
  { query: "Who attended the March 10 board meeting?", expectsContradiction: false },
];

function percentile(values: number[], p: number): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const idx = Math.min(sorted.length - 1, Math.floor(p * sorted.length));
  return sorted[idx];
}

function mean(values: number[]): number {
  if (values.length === 0) return 0;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function round(n: number): number {
  return Math.round(n * 100) / 100;
}

function peakRamMb(): number {
  return round(process.memoryUsage().rss / (1024 * 1024));
}

async function main() {
  const assertMode = process.argv.includes("--assert");
  console.log("=".repeat(64));
  console.log("  Quorum — Real Benchmark (QVAC SDK)");
  console.log("  Mode:", assertMode ? "ASSERT (CI gate)" : "REPORT");
  console.log("=".repeat(64));

  const totalRamGb = round(os.totalmem() / 1024 ** 3);
  console.log(`\n  Hardware: ${os.cpus()[0]?.model ?? os.arch()} | ${totalRamGb} GB RAM | ${os.cpus().length} cores`);
  console.log(`  Platform: ${os.platform()} ${os.release()} (${os.arch()})`);

  // ── Ingest corpus ──────────────────────────────────────────────────────────
  const files = readdirSync(DOSSIER_DIR).filter((f) => f.endsWith(".txt"));
  const documents = files.map((f) => `[Source: ${f}]\n${readFileSync(join(DOSSIER_DIR, f), "utf-8")}`);
  console.log(`\n  Ingesting ${files.length} documents...`);
  const ingestStart = performance.now();
  await resetCorpus(); // clean slate so latency/recall aren't skewed by stale duplicates
  await ingestCorpus(documents, true);
  const ingestMs = round(performance.now() - ingestStart);
  console.log(`  Corpus ingested in ${ingestMs}ms`);

  // ── Measure model load (cold) ────────────────────────────────────────────
  console.log("  Loading LLM...");
  const loadStart = performance.now();
  await loadLLMModel(LLAMA_MODEL_ID);
  const modelLoadMs = round(performance.now() - loadStart);
  console.log(`  Model loaded in ${modelLoadMs}ms`);

  // ── Run council over each query ──────────────────────────────────────────
  console.log(`\n  Running ${QUERIES.length} council queries...\n`);
  const results: Array<{
    query: string;
    total_ms: number;
    confidence: string;
    citations: number;
    expectsContradiction: boolean;
    contradictionCaught: boolean;
  }> = [];

  for (const { query, expectsContradiction } of QUERIES) {
    const t0 = performance.now();
    const council = await runQuorumCouncil(query);
    const totalMs = round(performance.now() - t0);
    // Per the PRD, the metric is whether the SKEPTIC catches the contradiction —
    // independent of whether the Synthesizer later downgrades overall confidence.
    const skepticTurn = council.turns.find((t) => t.role === "skeptic");
    const skepticCaught = skepticTurn ? skepticConcernLevel(skepticTurn.content) !== "none" : false;
    const contradictionCaught = skepticCaught || council.confidence !== "high";
    results.push({
      query,
      total_ms: totalMs,
      confidence: council.confidence,
      citations: council.citations.length,
      expectsContradiction,
      contradictionCaught,
    });
    console.log(
      `  ${totalMs.toFixed(0).padStart(7)}ms  ${council.confidence.padEnd(6)}  ` +
        `${council.citations.length} cites  ${query}`
    );
  }

  await releaseEmbeddingModel();

  // ── Aggregate ──────────────────────────────────────────────────────────────
  const totals = results.map((r) => r.total_ms);
  const planted = results.filter((r) => r.expectsContradiction);
  const caught = planted.filter((r) => r.contradictionCaught);
  const contradictionRecall = planted.length ? round(caught.length / planted.length) : null;
  const withCitations = results.filter((r) => r.citations > 0);
  const citationCoverage = results.length ? round(withCitations.length / results.length) : 0;
  const peak = peakRamMb();

  const stats = {
    council_p50_ms: round(percentile(totals, 0.5)),
    council_p95_ms: round(percentile(totals, 0.95)),
    council_mean_ms: round(mean(totals)),
    model_load_ms: modelLoadMs,
    ingest_ms: ingestMs,
    peak_ram_mb: peak,
    queries_run: results.length,
    contradiction_recall: contradictionRecall,
    contradiction_caught: `${caught.length}/${planted.length}`,
    citation_coverage: citationCoverage,
  };

  console.log("\n  ── Summary ──");
  console.log(`  Council p50: ${stats.council_p50_ms}ms | p95: ${stats.council_p95_ms}ms | mean: ${stats.council_mean_ms}ms`);
  console.log(`  Model load: ${stats.model_load_ms}ms | Ingest: ${stats.ingest_ms}ms | Peak RAM: ${peak} MB`);
  console.log(`  Contradiction recall: ${contradictionRecall} (${stats.contradiction_caught}) | Citation coverage: ${citationCoverage}`);

  // ── Write report ─────────────────────────────────────────────────────────
  const report = {
    timestamp: new Date().toISOString(),
    system: {
      platform: `${os.platform()}-${os.release()}-${os.arch()}`,
      processor: os.cpus()[0]?.model ?? os.arch(),
      cpu_count: os.cpus().length,
      ram_gb: totalRamGb,
      node: process.version,
    },
    budget: BUDGET,
    stats,
    queries: results,
    note: "Real @qvac/sdk measurements on target hardware",
  };
  const outDir = join(ROOT, "data");
  mkdirSync(outDir, { recursive: true });
  writeFileSync(join(outDir, "bench_results.json"), JSON.stringify(report, null, 2));
  console.log("\n  Results saved to data/bench_results.json");

  // ── Assert mode ────────────────────────────────────────────────────────────
  if (assertMode) {
    const failures: string[] = [];
    if (stats.council_p50_ms > BUDGET.council_total_ms)
      failures.push(`council_p50 ${stats.council_p50_ms}ms > budget ${BUDGET.council_total_ms}ms`);
    if (stats.model_load_ms > BUDGET.model_load_ms)
      failures.push(`model_load ${stats.model_load_ms}ms > budget ${BUDGET.model_load_ms}ms`);
    if (peak > BUDGET.peak_ram_mb) failures.push(`peak_ram ${peak}MB > budget ${BUDGET.peak_ram_mb}MB`);
    // Contradiction recall is reported, not gated: the 1B model is
    // non-deterministic, so a hard recall gate would be flaky. Surface it.
    if (contradictionRecall !== null && contradictionRecall < 1)
      console.log(`\n  ⚠️  contradiction_recall ${contradictionRecall} (${stats.contradiction_caught}) — informational; 1B model is non-deterministic run-to-run.`);
    if (failures.length) {
      console.log("\n  ❌ REGRESSION DETECTED:");
      failures.forEach((f) => console.log(`    • ${f}`));
      process.exit(1);
    }
    console.log("\n  ✅ All benchmarks within budget.");
  }

  console.log("\n" + "=".repeat(64));
}

main()
  .then(async () => {
    await close().catch(() => {}); // tear down the SDK worker so the process exits cleanly
    process.exit(0);
  })
  .catch(async (err) => {
    console.error("\n❌ Benchmark failed (model unavailable or SDK error):", err);
    console.error("   No results written — run with QVAC models available for real numbers.");
    await close().catch(() => {});
    process.exit(1);
  });
